src/types.py

```diff
@@ -323,3 +323,59 @@ class BuiltinMacro:
 
     function: Callable
     """The function to call to run the macro."""
+
+
+class TilingMode(IntEnum):
+    CUSTOM = -2
+    NONE = -1
+    DIRECTIONAL = 0
+    TILING = 1
+    CHARACTER = 2
+    ANIMATED_DIRECTIONAL = 3
+    ANIMATED = 4
+    STATIC_CHARACTER = 5
+    DIAGONAL_TILING = 6
+
+    def __str__(self) -> str:
+        if self == TilingMode.CUSTOM: return "custom"
+        if self == TilingMode.NONE: return "none"
+        if self == TilingMode.DIRECTIONAL: return "directional"
+        if self == TilingMode.TILING: return "tiling"
+        if self == TilingMode.CHARACTER: return "character"
+        if self == TilingMode.ANIMATED_DIRECTIONAL: return "animated_directional"
+        if self == TilingMode.ANIMATED: return "animated"
+        if self == TilingMode.STATIC_CHARACTER: return "static_character"
+        if self == TilingMode.DIAGONAL_TILING: return "diagonal_tiling"
```

Comment:
Still not too fussed by if chaining but you do you

Reply:
match is about the same level of performance (actually marginally worse) and adds an extra indentation level

Reply:
This is a hard violation of DRY though, to an unreasonable degree
I know DRY isn't enforced here but this is getting out of hand